### DISCUSSION TITLE
SIM800l

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1318] [Q&A] [answered]
I want to program the sim800l, but it didn't work. I was able to connect to the sim800l module with device.io.Serial, but I can't get the information I want, for example, I want to send a message, and when I send commands like this:

```js
this.#serialWrite("AT+CMGF=1"); // Configuring TEXT mode
// this.#serialWrite("AT+CSCS=\"HEX\"");
this.#serialWrite("AT+CSMP=49,167,0,8");
this.#serialWrite("AT+CMGS=\"+0000\"");
this.#serialWrite("Hi from moddable"); //text content
this.#serialWrite(26);
```

This way it displays the serial in the onReadable function, but I expect the response to be OK or ERROR, in Arduino everything works great, so the module is ok, but here I couldn't get the correct response:

AT+CMGF=1AT+CSMP=49,167,0,8AT+CMGS=\"+00000\" Hello from moddable

Full code :

```js
import Timer from "timer"
import TextDecoder from "text/decoder"

class SIM800 {

    #config = {}
    #serial = null

    constructor (config = {}) {
        this.#config = {
            ...config
        }

        this.#serial = new device.io.Serial({
            ...device.Serial.default,
            port: 2,
            baud: 9600,
            receive : 16,
            transmit: 17,
            ...config?.serial,
            format : "buffer",
            flowControl: "hardware",
            target : this,
            onReadable: this.#serialReadable
        });

        this.#serialWrite("AT+CMGF=1"); // Configuring TEXT mode
        // this.#serialWrite("AT+CSCS=\"HEX\"");
        this.#serialWrite("AT+CSMP=49,167,0,8");
        this.#serialWrite("AT+CMGS=\"+0000\"");
        this.#serialWrite("Hi from moddable"); //text content
        this.#serialWrite(26);
    }

    #serialReadable (count) {
        if ( count ) {
            const read = this.read()
            if ( read && read.byteLength ) {
                const decoder = new TextDecoder
                const string  = decoder.decode(read, {stream: true})
                if ( typeof string === "string" ) {
                    trace(`\nSIM800l : ${JSON.stringify(string)}`)
                }
            }
        }
    }

    #serialWrite (cmd = "") {
        Timer.delay(500)
        this.#serial.write(ArrayBuffer.fromString(cmd))
    }

}

export default SIM800
```

> Note: I'm sure the ESP32 is fully connected to the module, but it can't get the response

Maybe this arduino code will be more helpful for guidance :
```c++
#include <Arduino.h>
//#include <SoftwareSerial.h>                                // Remove if using HardwareSerial

#define RX_PIN 16
#define TX_PIN 17
#define BAUDRATE 9600 

HardwareSerial mySerial(2);                              // ESP32 Example 

void setup()
{
  //Begin serial communication with Arduino and Arduino IDE (Serial Monitor)
  Serial.begin(9600);
  
  //Begin serial communication with Arduino and SIM800L
  mySerial.begin(9600);

  Serial.println("Initializing..."); 
  delay(1000);

  mySerial.println("AT"); //Once the handshake test is successful, it will back to OK
  updateSerial();

  mySerial.println("AT+CMGF=1"); // Configuring TEXT mode
  updateSerial();
  //mySerial.println("AT+CSCS=\"HEX\"");
  //updateSerial();
  mySerial.println("AT+CSMP=49,167,0,8");
  updateSerial();
  mySerial.println("AT+CMGS=\"+0000\""); //country code, phone number to sms
  updateSerial();
  mySerial.print("Hi"); //text content
  updateSerial();
  mySerial.write(26);
}

void loop(){}

void updateSerial()
{
  delay(500);
  while (Serial.available()) 
  {
    mySerial.write(Serial.read());//Forward what Serial received to Software Serial Port
  }
  while(mySerial.available()) 
  {
    Serial.write(mySerial.read());//Forward what Software Serial received to Serial Port
  }
}
```

**Comment by salarizadi** [comment]
I tried to use this serial model but it didn't work because it doesn't work with the new version of Moddable SDK and it gives a lot of errors. Please check and update.

https://github.com/Moddable-OpenSource/moddable/tree/public/contributed/serial

**Reply by phoddie** [reply]
That module is not actively supported and may indeed be incompatible with ESP-IDF 5.x. We recommend using ECMA-419 IO (e.g. `new device.io.Serial`) as you do above.

**Reply by phoddie** [reply, marked as answer]
The Arduino example uses `println` which, I believe, adds a line feed at the end of the transmission. Your `#serialWrite` function uses `write()` which does not add a line feed. You probably need to do something like this:

```js
    #serialWrite (cmd = "") {
        Timer.delay(500)
        this.#serial.write(ArrayBuffer.fromString(cmd + "\n"));
    }
```

Also, this may not do what you intend:

```js
this.#serialWrite(26);
```

Your code will call `ArrayBuffer.fromString` which will output the characters "2" and "6". If you intend to output an ASCII 26, then you could do:

```js
this.#serial.write(Uint8Array.of(26));
```

**Reply by salarizadi** [reply]
@phoddie Thank you very much, I was able to send the message successfully, but I have a new problem. Unfortunately, when I try to send a long string, I get a full output error.

D:\ModdableSDK\modules\io\serial\esp32\serial.c (338) # Break: writing: full output!

In order for this module to support Farsi language, we need to send the message in HEX format, so first we convert the message, then we send the message using the following command:

```js
this.#serialWrite("06330644062706450020062706cc0646002006cc06a90020067e06cc06270645002006270632002000730069006d003800300030006c0020064506cc062806270634062f")
```

**Reply by phoddie** [reply]
ECMA-419 IO is primarily designed to be used asynchronously. The `onWritable` callback, in that case, gives you the number of bytes available to be written. You can use that to break up your writes so that they do not overflow the serial write buffer. If you want to operate purely synchronously, a quick hack here is to assume that the output buffer size (128 bytes on ESP32, I believe) and wait for it to empty (which at 9600 baud is about 1 ms per character).

```js
    #serialWrite (cmd = "") {
		cmd = new Uint8Array(ArrayBuffer.fromString(cmd + "\n"));
		const bufferLimit = 128;
		
		for (let position = 0; position < cmd.length; ) {
			const use = Math.min(cmd.length - position, bufferLimit);
			this.#serial.write(cmd.subarray(position, position + use));
			position += use;
			Timer.delay(use / 10);	// wait for write buffer to drain (calculation need adjusting)
		}
    }
```

**Reply by salarizadi** [reply]
@phoddie I tested this code but unfortunately I still get the same output full error, If you have another way, please tell me thank you

**Reply by phoddie** [reply]
> If you have another way..

Just because something doesn't work the first time, doesn't mean that you should immediately consider the approach fatally flawed and seek an alternative. I'm confident that the approach taken by the code above can work. It may need a little debugging, for example the delay calculation. I don't have the SIM800l to test with, so you will have to do the debugging yourself. For example, you could try doubling the delay to allow more time for the output buffer to flush. You could also try reducing the bufferLimit to 64, to stay further from the hardware buffer limit.

**Reply by salarizadi** [reply]
@phoddie Thank you so much for your tips, you are really great. Everything works properly. I will soon make this module available to other Moddable SDK developers. 🙏
